File: C++/169_Majority_Element.cpp
// https://leetcode.com/problems/majority-element/description/
// Boyer–Moore majority vote algorithm
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int vote=0;
        int res;
        for(auto n:nums){
            if (vote==0){
                res=n;
            }
            vote += (n==res)?1:-1;
        }
        return res;
    }
};

File: python/169_Majority_Element.py
# https://leetcode.com/problems/majority-element/description/
# Boyer–Moore majority vote algorithm
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        vote = 1
        res = nums[0]
        for i in range(1,len(nums)):
            if (nums[i]!=res):
                if (vote>0):
                    vote -= 1
                else:
                    vote=1
                    res = nums[i]
            else:
                vote += 1
        return res